In [1]:
import pandas as pd

customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")
products = pd.read_csv("../data/raw/olist_products_dataset.csv")
payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")
reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")
sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")
translation = pd.read_csv("../data/raw/product_category_name_translation.csv")

print("All main datasets loaded successfully.")

All main datasets loaded successfully.


In [2]:
print("Customers duplicates:", customers.duplicated().sum())
print("Orders duplicates:", orders.duplicated().sum())
print("Order items duplicates:", order_items.duplicated().sum())
print("Products duplicates:", products.duplicated().sum())
print("Payments duplicates:", payments.duplicated().sum())
print("Reviews duplicates:", reviews.duplicated().sum())
print("Sellers duplicates:", sellers.duplicated().sum())

Customers duplicates: 0
Orders duplicates: 0
Order items duplicates: 0
Products duplicates: 0
Payments duplicates: 0
Reviews duplicates: 0
Sellers duplicates: 0


In [4]:
orders["order_id"].duplicated().sum()

np.int64(0)

In [5]:
for name, df in {
    "customers": customers,
    "orders": orders,
    "order_items": order_items,
    "products": products,
    "payments": payments,
    "reviews": reviews,
    "sellers": sellers,
    "translation": translation
}.items():
    print(f"\n{name}")
    print(df.isna().sum()[df.isna().sum() > 0])


customers
Series([], dtype: int64)

orders
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
dtype: int64

order_items
Series([], dtype: int64)

products
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

payments
Series([], dtype: int64)

reviews
review_comment_title      87656
review_comment_message    58247
dtype: int64

sellers
Series([], dtype: int64)

translation
Series([], dtype: int64)


In [7]:
orders[orders["order_delivered_customer_date"].isna()][
    ["order_id", "order_status", "order_purchase_timestamp", "order_delivered_customer_date"]
].head(20)

,order_id,order_status,order_purchase_timestamp,order_delivered_customer_date
6,136cce7faa42fdb2cefd53fdc79a6098,invoiced,2017-04-11 12:22:08,NaN
44,ee64d42b8cf066f35eac1cf57de1aa85,shipped,2018-06-04 16:44:48,NaN
103,0760a852e4e9d89eb77bf631eaaf1c84,invoiced,2018-08-03 17:44:42,NaN
128,15bed8e2fec7fdbadb186b57c46c92f2,processing,2017-09-03 14:22:03,NaN
154,6942b8da583c2f9957e990d028607019,shipped,2018-01-10 11:33:07,NaN
162,36530871a5e80138db53bcfd8a104d90,shipped,2017-05-09 11:48:37,NaN
231,4d630f57194f5aba1a3d12ce23e71cd9,shipped,2017-11-17 19:53:21,NaN
266,8e24261a7e58791d10cb1bf9da94df5c,unavailable,2017-11-16 15:09:28,NaN
299,3b4ad687e7e5190db827e1ae5a8989dd,shipped,2018-06-28 12:52:15,NaN
305,b68d69564a79dea4776afa33d1d2fcab,shipped,2018-02-28 08:57:03,NaN


In [8]:
orders[orders["order_delivered_customer_date"].isna()]["order_status"].value_counts()

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

In [9]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [10]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [11]:
orders[orders["order_status"] == "delivered"]["order_delivered_customer_date"].isna().sum()

np.int64(8)

In [12]:
missing_delivered = orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_customer_date"].isna())
].shape[0]

print("Delivered orders missing delivery date:", missing_delivered)

Delivered orders missing delivery date: 8


In [13]:
orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_customer_date"].isna())
][[
    "order_id",
    "order_status",
    "order_purchase_timestamp",
    "order_delivered_carrier_date",
    "order_estimated_delivery_date"
]]

,order_id,order_status,order_purchase_timestamp,order_delivered_carrier_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,delivered,2017-11-28 17:44:07,2017-11-30 18:12:23,2017-12-18 00:00:00
20618,f5dd62b788049ad9fc0526e3ad11a097,delivered,2018-06-20 06:58:43,2018-06-25 08:05:00,2018-07-16 00:00:00
43834,2ebdfc4f15f23b91474edf87475f108e,delivered,2018-07-01 17:05:11,2018-07-03 13:57:00,2018-07-30 00:00:00
79263,e69f75a717d64fc5ecdfae42b2e8e086,delivered,2018-07-01 22:05:55,2018-07-03 13:57:00,2018-07-30 00:00:00
82868,0d3268bad9b086af767785e3f0fc0133,delivered,2018-07-01 21:14:02,2018-07-03 09:28:00,2018-07-24 00:00:00
92643,2d858f451373b04fb5c984a1cc2defaf,delivered,2017-05-25 23:22:43,NaN,2017-06-23 00:00:00
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,delivered,2018-06-08 12:09:39,2018-06-12 14:10:00,2018-06-26 00:00:00
98038,20edc82cf5400ce95e1afacc25798b31,delivered,2018-06-27 16:09:12,2018-07-03 19:26:00,2018-07-19 00:00:00


In [14]:
orders[
    [
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
].dtypes

order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

In [15]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for column in date_columns:
    orders[column] = pd.to_datetime(orders[column])

In [16]:
orders[date_columns].dtypes

order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

In [17]:
orders["delivery_days"] = (orders["order_delivered_customer_date"] - orders["order_purchase_timestamp"]).dt.days

In [18]:
orders[[
    "order_id",
    "order_purchase_timestamp",
    "order_delivered_customer_date",
    "delivery_days"
]].head(10)

,order_id,order_purchase_timestamp,order_delivered_customer_date,delivery_days
0,e481f51cbdc54678b7cc49136f2d6af7,2017-10-02 10:56:33,2017-10-10 21:25:13,8.0
1,53cdb2fc8bc7dce0b6741e2150273451,2018-07-24 20:41:37,2018-08-07 15:27:45,13.0
2,47770eb9100c2d0c44946d9cf07ec65d,2018-08-08 08:38:49,2018-08-17 18:06:29,9.0
3,949d5b44dbf5de918fe9c16f97b45f8a,2017-11-18 19:28:06,2017-12-02 00:28:42,13.0
4,ad21c59c0840e6cb83a9ceb5573f8159,2018-02-13 21:18:39,2018-02-16 18:17:02,2.0
5,a4591c265e18cb1dcee52889e2d8acc3,2017-07-09 21:57:05,2017-07-26 10:57:55,16.0
6,136cce7faa42fdb2cefd53fdc79a6098,2017-04-11 12:22:08,NaT,NaN
7,6514b8ad8028c9f2cc2374ded245783f,2017-05-16 13:10:30,2017-05-26 12:55:51,9.0
8,76c6e866289321a7c93b82b54852dc33,2017-01-23 18:29:09,2017-02-02 14:08:10,9.0
9,e69bfb5eb88e0ed6a785585b27e16dbf,2017-07-29 11:55:02,2017-08-16 17:14:30,18.0


In [19]:
orders["delivery_days"].describe()

count    96476.000000
mean        12.094086
std          9.551746
min          0.000000
25%          6.000000
50%         10.000000
75%         15.000000
max        209.000000
Name: delivery_days, dtype: float64

In [20]:
orders["delivery_days"].isna().sum()

np.int64(2965)

In [21]:
(orders["delivery_days"] < 0).sum()

np.int64(0)

In [22]:
orders["delivery_days"].value_counts().sort_index().head(20)

delivery_days
0.0       13
1.0     1572
2.0     3168
3.0     3849
4.0     4828
5.0     5810
6.0     6805
7.0     7653
8.0     6745
9.0     6039
10.0    5616
11.0    5114
12.0    4673
13.0    4376
14.0    3835
15.0    3170
16.0    2633
17.0    2360
18.0    2175
19.0    1816
Name: count, dtype: int64

In [23]:
orders["delivery_days"].value_counts().sort_index().head(20)

delivery_days
0.0       13
1.0     1572
2.0     3168
3.0     3849
4.0     4828
5.0     5810
6.0     6805
7.0     7653
8.0     6745
9.0     6039
10.0    5616
11.0    5114
12.0    4673
13.0    4376
14.0    3835
15.0    3170
16.0    2633
17.0    2360
18.0    2175
19.0    1816
Name: count, dtype: int64

In [24]:
orders["delivery_days"].value_counts().sort_index(ascending=False).head(10)

delivery_days
209.0    1
208.0    1
195.0    1
194.0    3
191.0    1
189.0    1
188.0    1
187.0    2
186.0    1
182.0    1
Name: count, dtype: int64

In [25]:
orders["delivery_delay_days"] = (
    orders["order_delivered_customer_date"]
    - orders["order_estimated_delivery_date"]
).dt.days

In [26]:
orders[[
    "order_id",
    "order_delivered_customer_date",
    "order_estimated_delivery_date",
    "delivery_delay_days"
]].head(10)

,order_id,order_delivered_customer_date,order_estimated_delivery_date,delivery_delay_days
0,e481f51cbdc54678b7cc49136f2d6af7,2017-10-10 21:25:13,2017-10-18,-8.0
1,53cdb2fc8bc7dce0b6741e2150273451,2018-08-07 15:27:45,2018-08-13,-6.0
2,47770eb9100c2d0c44946d9cf07ec65d,2018-08-17 18:06:29,2018-09-04,-18.0
3,949d5b44dbf5de918fe9c16f97b45f8a,2017-12-02 00:28:42,2017-12-15,-13.0
4,ad21c59c0840e6cb83a9ceb5573f8159,2018-02-16 18:17:02,2018-02-26,-10.0
5,a4591c265e18cb1dcee52889e2d8acc3,2017-07-26 10:57:55,2017-08-01,-6.0
6,136cce7faa42fdb2cefd53fdc79a6098,NaT,2017-05-09,NaN
7,6514b8ad8028c9f2cc2374ded245783f,2017-05-26 12:55:51,2017-06-07,-12.0
8,76c6e866289321a7c93b82b54852dc33,2017-02-02 14:08:10,2017-03-06,-32.0
9,e69bfb5eb88e0ed6a785585b27e16dbf,2017-08-16 17:14:30,2017-08-23,-7.0


In [27]:
orders["delivery_delay_days"].describe()

count    96476.000000
mean       -11.876881
std         10.183854
min       -147.000000
25%        -17.000000
50%        -12.000000
75%         -7.000000
max        188.000000
Name: delivery_delay_days, dtype: float64

In [28]:
print("Early:", (orders["delivery_delay_days"] < 0).sum())
print("On time:", (orders["delivery_delay_days"] == 0).sum())
print("Late:", (orders["delivery_delay_days"] > 0).sum())

Early: 88649
On time: 1292
Late: 6535


In [29]:
total_with_delay = orders["delivery_delay_days"].notna().sum()

early_pct = (orders["delivery_delay_days"] < 0).sum() / total_with_delay * 100
on_time_pct = (orders["delivery_delay_days"] == 0).sum() / total_with_delay * 100
late_pct = (orders["delivery_delay_days"] > 0).sum() / total_with_delay * 100

print(f"Early: {early_pct:.2f}%")
print(f"On time: {on_time_pct:.2f}%")
print(f"Late: {late_pct:.2f}%")

Early: 91.89%
On time: 1.34%
Late: 6.77%


### Delivery Performance

We calculated delivery delay by comparing the actual customer delivery date
with the estimated delivery date.

Among orders with both dates available:
- 91.89% arrived early
- 1.34% arrived on the estimated date
- 6.77% arrived late

Missing delivery dates were retained rather than artificially filled.

In [30]:
products[products["product_category_name"].isna()].head(10)

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
105,a41e356c76fab66334f36de622ecbd3a,NaN,NaN,NaN,NaN,650.0,17.0,14.0,12.0
128,d8dee61c2034d6d075997acef1870e9b,NaN,NaN,NaN,NaN,300.0,16.0,7.0,20.0
145,56139431d72cd51f19eb9f7dae4d1617,NaN,NaN,NaN,NaN,200.0,20.0,20.0,20.0
154,46b48281eb6d663ced748f324108c733,NaN,NaN,NaN,NaN,18500.0,41.0,30.0,41.0
197,5fb61f482620cb672f5e586bb132eae9,NaN,NaN,NaN,NaN,300.0,35.0,7.0,12.0
244,e10758160da97891c2fdcbc35f0f031d,NaN,NaN,NaN,NaN,2200.0,16.0,2.0,11.0
294,39e3b9b12cd0bf8ee681bbc1c130feb5,NaN,NaN,NaN,NaN,300.0,16.0,7.0,11.0
299,794de06c32a626a5692ff50e4985d36f,NaN,NaN,NaN,NaN,300.0,18.0,8.0,14.0
347,7af3e2da474486a3519b0cba9dea8ad9,NaN,NaN,NaN,NaN,200.0,22.0,14.0,14.0
428,629beb8e7317703dcc5f35b5463fd20e,NaN,NaN,NaN,NaN,1400.0,25.0,25.0,25.0


In [31]:
products[products["product_category_name"].isna()].shape

(610, 9)

In [32]:
missing_category_ids = products.loc[
    products["product_category_name"].isna(),
    "product_id"
]

sold_missing_category = order_items[
    order_items["product_id"].isin(missing_category_ids)
]

print("Products with missing category:", len(missing_category_ids))
print("Order items involving them:", len(sold_missing_category))

Products with missing category: 610
Order items involving them: 1603


In [33]:
translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [34]:
translation.columns

Index(['product_category_name', 'product_category_name_english'], dtype='str')

In [35]:
products = products.merge(
    translation,
    on="product_category_name",
    how="left"
)

In [36]:
products[
    ["product_category_name", "product_category_name_english"]
].head(10)

,product_category_name,product_category_name_english
0,perfumaria,perfumery
1,artes,art
2,esporte_lazer,sports_leisure
3,bebes,baby
4,utilidades_domesticas,housewares
5,instrumentos_musicais,musical_instruments
6,cool_stuff,cool_stuff
7,moveis_decoracao,furniture_decor
8,eletrodomesticos,home_appliances
9,brinquedos,toys


In [37]:
products["product_category"] = products["product_category_name_english"].fillna("Unknown")

In [39]:
products["product_category"].value_counts().head(20)

product_category
bed_bath_table                     3029
sports_leisure                     2867
furniture_decor                    2657
health_beauty                      2444
housewares                         2335
auto                               1900
computers_accessories              1639
toys                               1411
watches_gifts                      1329
telephony                          1134
baby                                919
perfumery                           868
stationery                          849
fashion_bags_accessories            849
cool_stuff                          789
garden_tools                        753
pet_shop                            719
Unknown                             623
electronics                         517
construction_tools_construction     400
Name: count, dtype: int64

In [40]:
products[
    products["product_weight_g"].isna() |
    products["product_length_cm"].isna() |
    products["product_height_cm"].isna() |
    products["product_width_cm"].isna()
][[
    "product_id",
    "product_category",
    "product_weight_g",
    "product_length_cm",
    "product_height_cm",
    "product_width_cm"
]]

,product_id,product_category,product_weight_g,product_length_cm,product_height_cm,product_width_cm
8578,09ff539a621711667c43eba6a3bd8466,baby,NaN,NaN,NaN,NaN
18851,5eb564652db742ff8f28759cd8d2652a,Unknown,NaN,NaN,NaN,NaN


In [41]:
missing_dimension_ids = products.loc[
    products["product_weight_g"].isna(),
    "product_id"
]

print(
    order_items[
        order_items["product_id"].isin(missing_dimension_ids)
    ].shape[0]
)

18


In [42]:
reviews["review_score"].isna().sum()

np.int64(0)

In [43]:
reviews["review_score"].value_counts().sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

In [1]:
df.to_csv("../data/clean_olist.csv", index=False)

NameError: name 'df' is not defined